079-deprecation-and-removal-of-storage-overrides.md

```diff
@@ -0,0 +1,95 @@
+But the Kafka Node Pools features was designed with the idea of using different configurations for different sets of nodes from the beginning.
+Users using node pools can have multiple different node pools - each representing a subset of Kafka nodes - with different configuration.
+This includes a different storage class.
+It also does need the user to specify exact node IDs as the storage configuration will be used for all nodes belonging to a given node pool.
+So it simplifies the configuration.
+
+In addition to simplifying the configuration for the users, it will also allow us to streamline the code and the API.
+In particular, the _storage diffing_ needed to avoid undesired changes will be simplified.
+
+## Proposal
+
+The persistent-volume-claim storage overrides will be deprecated immediately in Strimzi 0.43.
+The will remain supported and used, but users would be encouraged to move away from them and use Kafka Node Pools instead.
+This also falls into the time period when we expect most users to migrate to node pools and to KRaft which provides a good opportunity to move to per-node-pool storage class instead of using the overrides.
+
+Later, in the Strimzi version where support for ZooKeeper-based Kafka clusters is removed, the support for the storage overrides will be dropped and the overrides will be ignored.
+The overrides will remain in the CRD API for backwards compatibility but will not be used by the operator code.
+Finally, when moving the new `v1` CRD API in the future, the fields will be completely removed.
+
+Kubernetes do not allow changing storage class for existing persistent volumes. 
+If any user doesn't migrate from the overrides in time, the existing Kafka nodes and persistent volumes will not be affected.
+Only when the PVC/PV is deleted or when new nodes will be added during a scale-up, the overrides will be ignored and the default storage class will be used.
+So the impact on existing users should be minimal.
+
+### Warnings
+
+With the deprecation of the storage overrides, the operator will be updated to issue warnings when the overrides are used.
+The warnings will be printed in regular logs from the Cluster Operator.
+And they will be also added to the conditions in the `Kafka` CR status.
+
+## Affected projects
+
+This proposal affects the Strimzi Cluster Operator only.
+
+## Backwards compatibility
+
+This proposal removes the support for storage overrides in Kafka clusters.
+It will impact all users using it and they will have to migrate to node pools with different storage classes.
+Kafka clusters not using the overrides and any other operands supported by Strimzi will not be impacted in any way.
+
+## Rejected alternatives
+
+### Different timeline
+
+The timeline between the deprecation of the overrides and the removal of their support is relatively short (likely 3-4 Strimzi versions).
+If we think this is a problem, we can choose a different timeline.
```

Comment:
Would the removal be covered by another Proposal, or is it part of this one? I guess I'm mostly interested in what the operator does when the override support is removed but the user has configured the overrides: Would that still just be "log a warning + add a condition", or would the behaviour then be different?

Reply:
> Would the removal be covered by another Proposal, or is it part of this one?

This proposal includes the timeline -> on lines 59-61. The proposed timeline is to remove it together with ZooKeeper support (as in order to get to KRaft, everyone will need to use node pools and have a good chance to migrate away from the storage overrides as well). The place you are commenting on here is a rejected alternative. I'm basically trying to say here that I considered other timeline when it might be removed and rejected them. 

> I guess I'm mostly interested in what the operator does when the override support is removed but the user has configured the overrides: Would that still just be "log a warning + add a condition", or would the behaviour then be different?

As described on lines 63-66. When we start ignoring the overrides, nothing will happen right away with the existing volumes / brokers because the storage class is immutable. So the existing storage will keep exist as it was. What will change is that the storage override will be ignored when a new volume is created. That can happen for example when you scale-up the cluster with prepared override for the next node ID or when you delete the PV/PVC and let Strimzi create a new one.

Once we remove the support for the overrides, the API has to remain unchanged until v1 API is introduced. So users will still get the warnings based on the API field deprecations.